compiler/sem/semstmts.nim

```diff
@@ -1648,28 +1650,29 @@ proc semFor(c: PContext, n: PNode; flags: TExprFlags): PNode =
   if isCallExpr and call[0].kind == nkSym and
       call[0].sym.magic in {mFields, mFieldPairs}:
     result = semForFields(c, n, call[0].sym.magic)
-  elif isCallExpr and isClosureIterator(call[0].typ.skipTypes(abstractInst)):
-    # first class iterator:
-    result = semForVars(c, n, flags)
-  elif not isCallExpr or call[0].kind != nkSym or
-      call[0].sym.kind != skIterator:
-    if n.len == 3:
-      n[^2] = implicitIterator(c, "items", n[^2])
-    elif n.len == 4:
-      n[^2] = implicitIterator(c, "pairs", n[^2])
-    else:
-      localReport(c.config, n[^2], reportSem(rsemForExpectsIterator))
-    result = semForVars(c, n, flags)
   else:
+    if isCallExpr and isClosureIterator(call[0].typ.skipTypes(abstractInst)):
+      # first class iterator:
+      discard
+    elif not isCallExpr or call[0].kind != nkSym or
+        call[0].sym.kind != skIterator:
+      if n.len == 3:
+        n[^2] = implicitIterator(c, "items", n[^2])
+      elif n.len == 4:
+        n[^2] = implicitIterator(c, "pairs", n[^2])
+      else:
+        n[^2] = c.config.newError(n[^2], PAstDiag(kind: adSemForExpectedIterator))
+      hasError = n[^2].isError or hasError
     result = semForVars(c, n, flags)
-  if result.kind == nkError:
-    discard # do nothing
+  if hasError or result.kind == nkError:
+    hasError = true
```

Comment:
Is this right?

If `hasError` is `true`, then we really don't need to do anything, and if `result` is an error, then we only need to avoid the branches following this one? Is my understanding correct?

Reply:
I think here is just avoiding go to `elif` branch like previous

Reply:
yes, but setting `hasError` when the `result` is already an error isn't right. The convention is to set `hasError` if it has an error, not if it is one.

Reply:
so just remain old code ?